File: tests/refs/quick_sort.py
"""Reference: quicksort (Lomuto with randomised pivot) returning a sorted copy."""

from __future__ import annotations
import random
from typing import List

_rng = random.Random(42)


def _partition(arr: List[int], low: int, high: int) -> int:
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def _qs(arr: List[int], low: int, high: int) -> None:
    if low >= high:
        return
    pivot_idx = _rng.randint(low, high)
    arr[pivot_idx], arr[high] = arr[high], arr[pivot_idx]
    p = _partition(arr, low, high)
    _qs(arr, low, p - 1)
    _qs(arr, p + 1, high)


def quickSort(arr: List[int]) -> List[int]:
    a = list(arr)
    if a:
        _qs(a, 0, len(a) - 1)
    return a

File: tests/refs/QuickSort.java
/*
 * Reference Java implementation for tests/cases/quick_sort.json.
 * Lomuto quicksort with a seeded RNG for deterministic pivot choice.
 */
import java.util.Random;

public class QuickSort {
    private static final Random RNG = new Random(42);

    private static int partition(long[] a, int low, int high) {
        long pivot = a[high];
        int i = low - 1;
        for (int j = low; j < high; ++j) {
            if (a[j] <= pivot) {
                i++;
                long t = a[i]; a[i] = a[j]; a[j] = t;
            }
        }
        long t = a[i + 1]; a[i + 1] = a[high]; a[high] = t;
        return i + 1;
    }

    private static void qs(long[] a, int low, int high) {
        if (low >= high) return;
        int pivotIdx = low + RNG.nextInt(high - low + 1);
        long t = a[pivotIdx]; a[pivotIdx] = a[high]; a[high] = t;
        int p = partition(a, low, high);
        qs(a, low, p - 1);
        qs(a, p + 1, high);
    }

    public static long[] quickSort(long[] arr) {
        long[] a = arr.clone();
        if (a.length > 0) qs(a, 0, a.length - 1);
        return a;
    }
}
